src/combinatorics/simple_graph/adj_matrix.lean

```diff
@@ -15,89 +16,247 @@ properties to computational properties of the matrix.
+  dot_product vec (G.adj_matrix α v) = ∑ u in G.neighbor_finset v, vec u :=
 by simp [neighbor_finset_eq_filter, dot_product, sum_filter, finset.sum_apply]
 
 @[simp]
-lemma adj_matrix_mul_vec_apply (v : α) (vec : α → R) :
-  ((G.adj_matrix R).mul_vec vec) v = ∑ u in G.neighbor_finset v, vec u :=
+lemma adj_matrix_mul_vec_apply [non_assoc_semiring α] (v : V) (vec : V → α) :
+  ((G.adj_matrix α).mul_vec vec) v = ∑ u in G.neighbor_finset v, vec u :=
 by rw [mul_vec, adj_matrix_dot_product]
 
 @[simp]
-lemma adj_matrix_vec_mul_apply (v : α) (vec : α → R) :
-  ((G.adj_matrix R).vec_mul vec) v = ∑ u in G.neighbor_finset v, vec u :=
+lemma adj_matrix_vec_mul_apply [non_assoc_semiring α] (v : V) (vec : V → α) :
+  ((G.adj_matrix α).vec_mul vec) v = ∑ u in G.neighbor_finset v, vec u :=
 begin
   rw [← dot_product_adj_matrix, vec_mul],
   refine congr rfl _, ext,
-  rw [← transpose_apply (adj_matrix R G) x v, transpose_adj_matrix],
+  rw [← transpose_apply (adj_matrix α G) x v, transpose_adj_matrix],
 end
 
 @[simp]
-lemma adj_matrix_mul_apply (M : matrix α α R) (v w : α) :
-  (G.adj_matrix R ⬝ M) v w = ∑ u in G.neighbor_finset v, M u w :=
+lemma adj_matrix_mul_apply [non_assoc_semiring α] (M : matrix V V α) (v w : V) :
+  (G.adj_matrix α ⬝ M) v w = ∑ u in G.neighbor_finset v, M u w :=
 by simp [mul_apply, neighbor_finset_eq_filter, sum_filter]
 
 @[simp]
-lemma mul_adj_matrix_apply (M : matrix α α R) (v w : α) :
-  (M ⬝ G.adj_matrix R) v w = ∑ u in G.neighbor_finset w, M v u :=
+lemma mul_adj_matrix_apply [non_assoc_semiring α] (M : matrix V V α) (v w : V) :
+  (M ⬝ G.adj_matrix α) v w = ∑ u in G.neighbor_finset w, M v u :=
 by simp [mul_apply, neighbor_finset_eq_filter, sum_filter, adj_comm]
 
-variable (R)
-theorem trace_adj_matrix : matrix.trace α R R (G.adj_matrix R) = 0 := by simp
-variable {R}
+variable (α)
+
+theorem trace_adj_matrix [non_assoc_semiring α] [semiring β] [module β α]:
+  matrix.trace _ β _ (G.adj_matrix α) = 0 :=
+by simp
+
+variable {α}
 
-theorem adj_matrix_mul_self_apply_self (i : α) :
-  ((G.adj_matrix R) ⬝ (G.adj_matrix R)) i i = degree G i :=
+theorem adj_matrix_mul_self_apply_self [non_assoc_semiring α] (i : V) :
+  ((G.adj_matrix α) ⬝ (G.adj_matrix α)) i i = degree G i :=
 by simp [degree]
 
 variable {G}
 
 @[simp]
-lemma adj_matrix_mul_vec_const_apply {r : R} {v : α} :
-  (G.adj_matrix R).mul_vec (function.const _ r) v = G.degree v * r :=
+lemma adj_matrix_mul_vec_const_apply [semiring α] {a : α} {v : V} :
+  (G.adj_matrix α).mul_vec (function.const _ a) v = G.degree v * a :=
 by simp [degree]
 
-lemma adj_matrix_mul_vec_const_apply_of_regular {d : ℕ} {r : R} (hd : G.is_regular_of_degree d)
-  {v : α} :
-  (G.adj_matrix R).mul_vec (function.const _ r) v = (d * r) :=
+lemma adj_matrix_mul_vec_const_apply_of_regular
+[semiring α] {d : ℕ} {a : α} (hd : G.is_regular_of_degree d)
+  {v : V} :
+  (G.adj_matrix α).mul_vec (function.const _ a) v = (d * a) :=
 by simp [hd v]
 
 end simple_graph
```

Comment:
Showing to_graph is a left inverse:
```lean
namespace matrix.is_adj_matrix
variables {A : matrix V V α} [mul_zero_one_class α] [nontrivial α] (h : is_adj_matrix A)

lemma adj_matrix_to_graph_eq [decidable_eq α] :
  h.to_graph.adj_matrix α = A :=
begin
  ext i j,
  obtain (h'|h') := h.zero_or_one i j; simp [h'],
end

end matrix.is_adj_matrix
```

Reply:
@l534zhan Is this still outstanding? If it's done, please mark this as resolved.

Reply:
@jcommelin Yes. Johan, what is the proper name for this? to_graph_adj_matrix_eq or adj_matrix_to_graph_eq?

Reply:
@l534zhan The names I suggested are consistent with the other ones you've introduced in the file (like `is_symm_compl` and `is_adj_matrix_adj_matrix`). Though `compl_to_graph_eq` should then be `to_graph_compl_eq`.